astropy/utils/data.py

```diff
@@ -1327,43 +1406,303 @@ def _get_download_cache_locs():
+    have accumulated in the cache. These can be removed with
+    `clear_download_cache`, either passing the filename returned here, or
+    with no arguments to empty the entire cache and return it to a
+    reasonable, if empty, state.
+
+    Parameters
+    ----------
+    check_hashes : bool, optional
+        Whether to compute the hashes of the contents of all files in the
+        cache and compare them to the names. This can take some time if
+        the cache contains large files.
+
+    Returns
+    -------
+    strays : set of strings
+        This is the set of files in the cache directory that do not correspond
+        to known URLs. This may include some files associated with the cache
+        index.
+
+    Raises
+    ------
+    CacheDamaged
+        To indicate a problem with the cache contents; the exception contains
+        a .bad_url or a .bad_file attribute to allow the user to use
+        `clear_download_cache` to remove the offending item.
+    OSError, RuntimeError
+        To indicate some problem with the cache structure. This may need a full
+        `clear_download_cache` to resolve, or may indicate some kind of
+        misconfiguration.
+    """
+    # We want full powers over the cache to carry out a proper check
+    with _cache() as (dldir, url2hash):
+        if dldir is None:
+            raise OSError("Cache directory cannot be created or accessed")
+        leftover_files = set(os.path.join(dldir, k)
+                             for k in os.listdir(dldir))
+        leftover_files.discard(os.path.join(dldir, "urlmap"))
+        leftover_files.discard(os.path.join(dldir, "lock"))
+        for u, h in url2hash.items():
+            if not os.path.exists(h):
+                msg = "URL '{}' points to nonexistent file '{}'".format(
+                        u, h)
+                raise CacheDamaged(msg, bad_url=u)
+            leftover_files.discard(h)
+            d, hexdigest = os.path.split(h)
+            if dldir != d:
+                msg = ("Expected downloaded files to be in '{}' but "
+                       "URL '{}' points to file '{}'"
+                       .format(dldir, u, h))
+                raise CacheDamaged(msg, bad_url=u)
+            if check_hashes:
+                hexdigest_file = compute_hash(h)
+                if hexdigest_file != hexdigest:
+                    msg = ("File corresponding to {} has contents that "
+                           "do not match the hash value: '{}' "
+                           .format(u, hexdigest))
+                    raise CacheDamaged(msg, bad_file=h)
+        for h in leftover_files:
+            h_base = os.path.basename(h)
+            if len(h_base) >= 32 and re.match("[0-9a-f]+", h_base):
+                raise CacheDamaged("Apparently abandoned hash file {}"
+                                   .format(h), bad_file=h)
+    return leftover_files
+
+
+def import_to_cache(url_key, filename,
```

Comment:
Should the name be a bit more descriptive? Or `insert_file_to_cache`?
```suggestion
def import_file_to_cache(url_key, filename,
```

Reply:
Good idea.